MANTA-4897 Overhaul rust zookeeper client MANTA-5215 Add multi-server support to rust zookeeper client MANTA-5217 Implement backoff stability threshold for rust zookeeper client

isaacdavis opened the pull request:
I'm putting this up now, even though there's more left to do. I'll file tickets for these things that are missing:
- [DONE] multiple-server connect string
- [DONE] stability threshold before resetting backoff
- test suite

the two "DONE" items are now a part of this pull request.

Commit:
MANTA-4897 Overhaul rust zookeeper client

Commit:
remove backoff.rs.old

kellymclaughlin commented on src/proto/decoder.rs:
```diff
@@ -0,0 +1,338 @@
+    // Global map of pending watches.
+    //
+    // Watches are only registered once we have confirmed that the operation
+    // that initiated the watch did not fail. Thust, we must stage watches here
```
typo: Thust

isaacdavis replied:
Fixed

kellymclaughlin commented on src/proto/decoder.rs:
```diff
@@ -0,0 +1,338 @@
+    type Error = IoError;
+
+    fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
+        fn wrap<T, E>(item: T) -> Result<Option<T>, E> {
```
I'd suggest moving this to be just a top-level helper function since it doesn't operate in terms of anything inside the `Decoder` implementation.

isaacdavis replied:
With some of the other refactoring I decided that this function wasn't really necessary.

kellymclaughlin commented on src/proto/decoder.rs:
```diff
@@ -0,0 +1,338 @@
+        //
+        // A watch has triggered.
+        //
+        if xid == WATCH_XID {
```
The flow of this gets a little hard to follow due to the size of this outer `if/elseif/else` block. It might be nice to break out the handling of events from the control flow. Something like:
```
if xid == WATCH_XID {
    handle_watch_event(...)
} else if xid == HEARTBEAT_XID {
   handle_heartbeat(...)
} else {
    handle_request(...)
}

isaacdavis replied:
This is very true. I pulled the branches out into functions (except for the heartbeat since it's just a few lines).

kellymclaughlin commented on src/proto/request.rs:
```diff
@@ -120,73 +133,82 @@ pub(super) enum MultiHeader {
-        writer.write_i32::<BigEndian>(self.len() as i32)?;
-        writer.write_all(self.as_ref())
+//
+// XXX is there a way to avoid implementing WriteTo for both String and &String?
```
I don't think there is since the reference is treated as a separate type.

isaacdavis replied:
Ah, well. I deleted the comment.

kellymclaughlin reviewed (commented):
I haven't made it through all of this in detail yet, but I've got a few comments while I continue to go through it. This is really nice stuff!

kellymclaughlin commented on src/client.rs:
```diff
@@ -0,0 +1,611 @@
+            sess_mgr: SessionManager,
+            log: Logger,
+        ) -> Result<(), InternalError> {
+            match decoder.next().await {
```
You could use an `if let` here to get rid of one level of nesting.

Commit:
feedback round and added multi-server capability

Commit:
more feedback

isaacdavis commented:
Addressed first round of feedback.

I've also added multi-server support with ZkConnectString, borrowing the code from the primary resolver. This is MANTA-5215.

Commit:
added backoff stability threshold

isaacdavis commented:
Added backoff stability threshold. This adds the file `backoff.rs` and slightly simplifies some of the code in `session_manager.rs`. See MANTA-5217 for testing notes for this piece of functionality.

kellymclaughlin commented on src/lib.rs:
````diff
@@ -93,283 +93,192 @@
-//! # A somewhat silly example
+//! # An example
 //!
 //! ```no_run
````
I get an error here when running `cargo test`. I think it should be `rust,ignore` instead of `no_run` for now to avoid that error.

kellymclaughlin commented on src/types/mod.rs:
```diff
@@ -1,3 +1,7 @@
+//
```
I think nowadays it's more idiomatic to have a `type.rs` file at the same directory level and the `types` directory and then only include the specific submodules inside the `types` directory.

kellymclaughlin commented on src/proto/mod.rs:
```diff
@@ -1,28 +1,8 @@
-use std::net::SocketAddr;
-use tokio;
-use tokio::prelude::*;
+//
```
Same as for my comment about the `types` module I would move this up one level to a `proto.rs` file.

kellymclaughlin commented on src/types/mod.rs:
```diff
@@ -76,3 +109,73 @@ pub enum CreateMode {
 }
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum ZkConnectStringError {
```
Add a comment for this type similar to the one above for `CreateMode` so that a description shows up in the generated docs.

kellymclaughlin reviewed (commented):
I found a few missing links in the crate-level documentation. The following changes address it:
```
diff --git a/src/lib.rs b/src/lib.rs
index cb7be98..4cc1038 100644
--- a/src/lib.rs
+++ b/src/lib.rs
@@ -41,9 +41,9 @@
 //!
 //! ## Ephemeral nodes
 //!
-//! When you create a Znode, you also specify a [`CreateMode`]. Nodes that are created with
-//! [`CreateMode::Persistent`] are the nodes we have discussed thus far. They remain in the server
-//! until you delete them. Nodes that are created with [`CreateMode::Ephemeral`] on the other hand
+//! When you create a Znode, you also specify a [`CreateMode`](types/enum.CreateMode.html). Nodes that are created with
+//! [`CreateMode::Persistent`](types/enum.CreateMode.html#variant.Persistent) are the nodes we have discussed thus far. They remain in the server
+//! until you delete them. Nodes that are created with [`CreateMode::Ephemeral`](types/enum.CreateMode.html#variant.Ephemeral) on the other hand
 //! are special. These [ephemeral
 //! nodes](https://zookeeper.apache.org/doc/current/zookeeperProgrammers.html#Ephemeral+Nodes) are
 //! automatically deleted by the server when the client that created them disconnects. This can be
@@ -53,8 +53,8 @@
 //!
 //! ## Watches
 //!
-//! In addition to the methods above, [`ZooKeeper::exists`], [`ZooKeeper::get_data`], and
-//! [`ZooKeeper::get_children`] also support setting
+//! In addition to the methods above, [`ZooKeeper::exists`](struct.ZooKeeper.html#method.exists), [`ZooKeeper::get_data`](struct.ZooKeeper.html#method.get_data), and
+//! [`ZooKeeper::get_children`](struct.ZooKeeper.html#method.get_children) also support setting
 //! "[watches](https://zookeeper.apache.org/doc/current/zookeeperProgrammers.html#ch_zkWatches)" on
 //! a node. A watch is one-time trigger that causes a [`WatchedEvent`] to be sent to the client
 //! that set the watch when the state for which the watch was set changes. For example, for a
```

isaacdavis commented:
Closing this PR; should it become relevant in the future please feel free to reopen and reassign as necessary.
